moabb/analysis/results.py

```diff
@@ -146,6 +142,189 @@ def _already_computed(self, pipeline, dataset, subject):
+class ResultsDB:
+    '''Class to interface with results database. Can add new data and also return
+    dataframes based on queries 
+
+    The integrity of this structure depends on the pipelines and the paradigms
+    having unambiguous __repr__ methods.
+
+    '''
+
+    def __init__(self, write=False, evaluation=None, _debug=False):
+        """
+        Initialize class. If database does not exist, create it.
+        write: bool, do you want to write values (if yes, then need eval and paradigm)
+        evaluation: BaseEvaluation child
+        _debug: flag to save to separate database for testing
+        """
+        # first ensure that the database exists
+        import moabb
+        self.mod_dir = os.path.dirname(
+            os.path.abspath(inspect.getsourcefile(moabb)))
+        if not _debug:
+            self.filepath = os.path.join(self.mod_dir, 'results', 'results.db')
+        else:
+            self.filepath = os.path.join(
+                self.mod_dir, 'results', 'results_test.db')
+        if not os.path.isfile(self.filepath):
+            os.makedirs(os.path.join(self.mod_dir, 'results'), exist_ok=True)
+            self._setup()
+        self.conn = sqlite3.connect(self.filepath)
+
+        self.write = write
+        if write:
+            if evaluation is None:
+                raise ValueError(
+                    'If writing results, evaluation must be specified')
+            self.evaluation = repr(evaluation)
+            self.paradigm = repr(evaluation.paradigm))
+            self.human_paradigm = evaluation.paradigm.human_paradigm
+            self.context_id = self.check_context(
+                get_digest(evaluation), get_digest(evaluation.paradigm),
+                self.human_paradigm)
+
+    def check_context(self, eval_hash, par_hash, human_par_hash):
+        '''
+        Check if evaluation, paradigm, and human paradigm combination exists already.
+        If so, return the generated ID. If not, insert and return the appropriate ID
+        '''
+        with self.conn as c:
+            id_list = c.execute("SELECT id FROM context WHERE eval = ? AND pprocess_hash = ? AND paradigm_hash = ?", (
+                eval_hash, par_hash, human_par_hash)).fetchall()
+            log.debug(id_list)
+            context_id = None
+            if len(id_list) == 0:
+                log.info('Adding new context to database')
+                cur = c.cursor()
+                cur.execute("INSERT INTO context(eval, pprocess_hash, paradigm_hash) VALUES(?,?,?)",
+                            (eval_hash, par_hash, human_par_hash))
+                context_id = cur.lastrowid
+            else:
+                context_id = id_list[0][0]
+            return context_id
+
+    def _setup(self):
+        '''
+        Set up initial table structure 
+        '''
+        self.conn = sqlite3.connect(self.filepath)
+        with self.conn as c:
+            c.execute('''
+            CREATE TABLE context(id INTEGER PRIMARY KEY,
```

Comment:
i would also create a table for paradigm and eval, then link them in a context ?

Reply:
Do we need a table for paradigm/eval? I think it's sufficient to save the repr since that has all relevant information